vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+	// key can safely be deleted now
+	err = req.Storage.Delete(ctx, namedKeyConfigPath+targetKeyName)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = i.expireOIDCPublicKeys(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	i.oidcCache.purge()
+
+	return nil, nil
+}
+
+// handleOIDCListKey is used to list named keys
+func (i *IdentityStore) pathOIDCListKey(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	keys, err := req.Storage.List(ctx, namedKeyConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(keys), nil
+}
+
+// pathOIDCRotateKey is used to manually trigger a rotation on the named key
+func (i *IdentityStore) pathOIDCRotateKey(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	// load the named key and perform a rotation
+	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return logical.ErrorResponse("no named key found at %q", name), logical.ErrInvalidRequest
+	}
+
+	var storedNamedKey namedKey
+	if err := entry.DecodeJSON(&storedNamedKey); err != nil {
+		return nil, err
+	}
+
+	// call rotate with an appropriate overrideTTL where < 0 means no override
+	verificationTTLOverride := -1 * time.Second
+
+	if ttlRaw, ok := d.GetOk("verification_ttl"); ok {
+		verificationTTLOverride = time.Duration(ttlRaw.(int)) * time.Second
+	}
+
+	if err := storedNamedKey.rotate(ctx, req.Storage, verificationTTLOverride); err != nil {
+		return nil, err
+	}
+
+	i.oidcCache.purge()
+
+	return nil, nil
+}
+
+func (i *IdentityStore) pathOIDCKeyExistenceCheck(ctx context.Context, req *logical.Request, d *framework.FieldData) (bool, error) {
+	name := d.Get("name").(string)
+
+	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+name)
+	if err != nil {
+		return false, err
+	}
+
+	return entry != nil, nil
+}
+
+// handleOIDCGenerateSignToken generates and signs an OIDC token
+func (i *IdentityStore) pathOIDCGenerateToken(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("name").(string)
+
+	role, err := i.getOIDCRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse("role %q not found", roleName), nil
+
+	}
+
+	entry, _ := req.Storage.Get(ctx, namedKeyConfigPath+role.Key)
+	if entry == nil {
+		return logical.ErrorResponse("key %q not found", role.Key), nil
+	}
+
+	var key namedKey
+	if err := entry.DecodeJSON(&key); err != nil {
+		return nil, err
+	}
+
+	// generate an OIDC token from entity data
+	accessorEntry, err := i.core.tokenStore.lookupByAccessor(ctx, req.ClientTokenAccessor, false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	te, err := i.core.LookupToken(ctx, accessorEntry.TokenID)
+	if err != nil {
+		return nil, err
+	}
+
+	if te == nil || te.EntityID == "" {
```

Comment:
OK so basically all of this lookup is to get a token entry. But we actually have it attached to the request and are simply nil-ing it out in the router. I think rather than go through all this you should update router such that if the backend that's being routed to is of type identity (and maybe ns-identity when we get to that point) then the token entry isn't nil'd.

Reply:
See above comment... I think `te` goes away.